micro_sam/sam_3d_wrapper.py

```diff
@@ -0,0 +1,151 @@
+        embed_dim: int = 768,
+        **kwargs
+    ):
+        super().__init__()
+        self.image_encoder = image_encoder
+        self.img_size = self.image_encoder.img_size
+
+        # freeze sam blocks
+        for k, v in self.image_encoder.named_parameters():
+            if '.adapter_' not in k:
+                v.requires_grad = False
+
+        # replace default blocks with 3d adapter blocks
+        for i, blk in enumerate(self.image_encoder.blocks):
+            self.image_encoder.blocks[i] = NDBlock(Block=blk, num_heads=num_heads, dim=embed_dim)
+
+    def forward(self, x: torch.Tensor, d_size) -> torch.Tensor:
+        x = self.image_encoder.patch_embed(x)
+        if self.image_encoder.pos_embed is not None:
+            x = x + self.image_encoder.pos_embed
+
+        for blk in self.image_encoder.blocks:
+            x = blk(x, d_size)
+
+        x = self.image_encoder.neck(x.permute(0, 3, 1, 2))
+
+        return x
+
+
+class NDBlock(nn.Module):
+    def __init__(
+            self,
+            dim: int,
+            num_heads: int,
+            Block: nn.Module,
+            norm_layer: Type[nn.Module] = nn.LayerNorm,
+            **kwargs
+    ):
+        super().__init__()
+        self.Block = Block
+        
+        self.adapter_channels = 384
+        self.adapter_linear_down = nn.Linear(dim, self.adapter_channels, bias=False)
+        self.adapter_linear_up = nn.Linear(self.adapter_channels, dim, bias=False)
+        self.adapter_conv = nn.Conv3d(self.adapter_channels, self.adapter_channels, kernel_size=(3, 1, 1), padding='same')
+        self.adapter_act = nn.GELU()
+        self.adapter_norm = norm_layer(dim)
+
+        self.adapter_linear_down_2 = nn.Linear(dim, self.adapter_channels, bias=False)
+        self.adapter_linear_up_2 = nn.Linear(self.adapter_channels, dim, bias=False)
+        self.adapter_conv_2 = nn.Conv3d(self.adapter_channels, self.adapter_channels, kernel_size=(3, 1, 1), padding='same')
+        self.adapter_act_2 = nn.GELU()
+        self.adapter_norm_2 = norm_layer(dim)
+
+    def forward(self, x: torch.Tensor, d_size) -> torch.Tensor:
+        b_size, hw_size = x.shape[0], x.shape[1]
+        
+        # 3D adapter
+        shortcut = x
+        x = self.adapter_norm(x)
+        x = self.adapter_linear_down(x)
+        x = x.contiguous().view(int(b_size/d_size), d_size, hw_size, hw_size, self.adapter_channels)
+        x = torch.permute(x, (0, -1, 1, 2, 3))
+        x = self.adapter_conv(x)
+        x = torch.permute(x, (0, 2, 3, 4, 1))
+        x = x.contiguous().view(b_size, hw_size, hw_size, self.adapter_channels)
+        x = self.adapter_act(x)
+        x = self.adapter_linear_up(x)
+        x = shortcut + x
+        # end 3D adapter
+        
+        shortcut = x
+        x = self.norm1(x)
+        # Window partition
+        if self.window_size > 0:
+            H, W = x.shape[1], x.shape[2]
+            x, pad_hw = window_partition(x, self.window_size)
+
+        x = self.attn(x)
```

Comment:
This is never defined! I think this has to be `self.Block.attn(x)` (note that I don't think that `Block` should be capitalized, so ideally `self.block.attn(x)` instead.)